### PR TITLE
Fix/revpi 2856/adapt telegrams to serdev

### PR DESCRIPTION
**Pull request by linosanfilippo-kunbus**

### CODE REVIEW

**Comment by l1k** [comment]
Nit: The commit message of [ca7f7e5](https://github.com/RevolutionPi/piControl/pull/86/commits/ca7f7e52b93769cfe7cd9be95d7b5bcf47a82534) should probably use the term "user telegrams" instead of "telegram communication" to signify this is only for userspace-provided telegrams, not any internally generated telegrams.

File: RevPiDevice.c [hunk 1]
```diff
@@ -243,7 +244,16 @@ int RevPiDevice_run(void)
...
+		char rcv_buff[IOPROTOCOL_MAXDATA_LENGTH];
+
+		piCore_g.statusUserTel = pibridge_req_io(hdr->sHeaderTyp1.bitAddress,
+							 hdr->sHeaderTyp1.bitCommand,
+							 telegram->ai8uData,
+							 hdr->sHeaderTyp1.bitLength,
+							 rcv_buff,
+							 sizeof(rcv_buff));
 		up(&piCore_g.semUserTel);
 	}
```
Comment (l1k) [inline RevPiDevice.c:258]: The `KB_INTERN_IO_MSG` ioctl expects the response telegram in `piCore_g.responseUserTel`, yet you're only temporarily storing the response in `rcv_buff` on the stack. The stack gets thrown away when exiting the function.

I'd suggest passing `&piCore_g.responseUserTel.ai8uData` and `IOPROTOCOL_MAX_DATA_LENGTH` (or alternatively `sizeof(piCore_g.responseUserTel.ai8uData) - 1`) to `pibridge_req_io()`.

File: piIOComm.c [hunk 2]
```diff
@@ -319,83 +319,6 @@ INT32S piIoComm_sendRS485Tel(INT16U i16uCmd_p, INT8U i8uAddress_p,
...
-		pr_info("dev %2d: send cyclic Data addr %d + %d output 0x%02x 0x%02x\n",
-			pRequest_p->uHeader.sHeaderTyp1.bitAddress,
-			pRequest_p->uHeader.sHeaderTyp1.bitAddress,
-			RevPiDevice.dev[i8uDevice_p].i16uOutputOffset, sRequest_l.ai8uData[0], sRequest_l.ai8uData[1]);
-	}
-#endif
-
-	ret = piIoComm_send((INT8U *) pRequest_p, IOPROTOCOL_HEADER_LENGTH + len_l + 1);
-	if (ret == 0) {
-		ret = piIoComm_recv((INT8U *) pResponse_p, REV_PI_RECV_IO_HEADER_LEN);
```
Comment (l1k) [inline piIOComm.c:350]: You're removing the last occurrence of `REV_PI_RECV_IO_HEADER_LEN` from the code, so please remove the macro from piIOComm.h as well.

The macro is defined as 65530 and I did some git archeology to find out why: It turns out that prior to the serdev migration, `UartThreadProc()` contained a <a href="https://github.com/RevolutionPi/piControl/blob/master/piIOComm.c#L129">special case</a> whereby the length of the received data is taken from the packet header (instead of actually trying to receive 65530 bytes). What an egregious hack! On the serdev branch, we eliminated `UartThreadProc()` and thus the special case, so I think on the serdev branch we're currently always running into the timeout when receiving a user telegram because we're waiting for 65530 bytes. Good thing that you're fixing it now!

File: RevPiDevice.c (hunk 1 above)
Comment (l1k) [reply RevPiDevice.c:258]: The `- 1` is for the CRC in case that wasn't clear.

File: RevPiDevice.c [hunk 3]
```diff
@@ -243,7 +244,16 @@ int RevPiDevice_run(void)
 
 	// if the user-ioctl want to send a telegram, do it now
 	if (test_bit(REVPI_CORE_FLAG_PENDING_TELEGRAM, &piCore_g.flags)) {
-		piCore_g.statusUserTel = piIoComm_sendTelegram(&piCore_g.requestUserTel, &piCore_g.responseUserTel);
+		SIOGeneric *telegram = &piCore_g.requestUserTel;
+		UIoProtocolHeader *hdr = &telegram->uHeader;
+		char rcv_buff[IOPROTOCOL_MAXDATA_LENGTH];
+
+		piCore_g.statusUserTel = pibridge_req_io(hdr->sHeaderTyp1.bitAddress,
```
Comment (l1k) [inline RevPiDevice.c:251]: This seems to exceed 80 chars. Maybe move the `pibridge_req_io(hdr->sHeaderTyp1.bitAddress,` to the next line and indent with 3 tabs? Can't really think of a better solution.

File: piIOComm.c [hunk 4]
```diff
@@ -311,7 +311,7 @@ INT32S piIoComm_sendRS485Tel(INT16U i16uCmd_p, INT8U i8uAddress_p,
 	ret = pibridge_req_gate_tmt(i8uAddress_p, i16uCmd_p, pi8uSendData_p,
 				    i8uSendDataLen_p, pi8uRecvData_p,
 				    rcvlen, timeout);
-	if (ret) {
+	if (ret < 0) {
```
Comment (l1k) [inline piIOComm.c:314]: I think you need to check for `ret != rcvlen` instead. Otherwise you'll happily accept telegrams which are shorter than expected.

File: piAIOComm.c [hunk 5]
```diff
@@ -268,23 +268,23 @@ u32 piAIOComm_Init(u8 devnum)
 	pr_info_aio("piAIOComm_Init send configIn1\n");
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA2, snd_buf,
 			      AIO_CONFIG_DATA2_LEN, NULL, 0);
-	if (ret)
+	if (ret < 0)
```
Comment (l1k) [inline piAIOComm.c:271]: I think you don't need to change anything here. You're expecting an empty response (payload with 0 bytes). So receiving anything greater than 0 would be an error.

File: piAIOComm.c [hunk 6]
```diff
@@ -268,23 +268,23 @@ u32 piAIOComm_Init(u8 devnum)
...
 		return 3;
 
 	snd_buf = &aioIn2Config_s[dev_idx];
 
 	pr_info_aio("piAIOComm_Init send configIn2\n");
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA3, snd_buf,
 			      AIO_CONFIG_DATA3_LEN, NULL, 0);
-	if (ret)
+	if (ret < 0)
```
Comment (l1k) [inline piAIOComm.c:279]: Same here, change is not necessary.

File: piAIOComm.c [hunk 7]
```diff
@@ -268,23 +268,23 @@ u32 piAIOComm_Init(u8 devnum)
...
 		return 3;
 
 	snd_buf = &aioConfig_s[dev_idx];
 
 	pr_info_aio("piAIOComm_Init send config\n");
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_CFG, snd_buf,
 			      AIO_CONFIG_DATA1_LEN, NULL, 0);
-	if (ret)
+	if (ret < 0)
```
Comment (l1k) [inline piAIOComm.c:287]: Same here, change is not necessary.

File: piAIOComm.c [hunk 8]
```diff
@@ -318,7 +318,7 @@ u32 piAIOComm_sendCyclicTelegram(u8 devnum)
 
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA, snd_buf,
 			      AIO_OUTPUT_DATA_LEN, rcv_buf, AIO_INPUT_DATA_LEN);
-	if (ret) {
+	if (ret < 0) {
```
Comment (l1k) [inline piAIOComm.c:321]: I think you need to check for `ret != AIO_INPUT_DATA_LEN` instead so that you treat too-short packets as an error (as before).

File: piDIOComm.c [hunk 9]
```diff
@@ -139,6 +139,8 @@ INT32U piDIOComm_Init(INT8U i8uDevice_p)
 
 			ret = pibridge_req_io(addr, IOP_TYP1_CMD_CFG, snd_buf,
 					      snd_len, NULL, 0);
+			if (ret > 0) /* success */
+				ret = 0;
```
Comment (l1k) [inline piDIOComm.c:143]: Change not necessary, you're expecting an empty payload as response, so anything greater than 0 is an error.

File: piDIOComm.c [hunk 10]
```diff
@@ -207,7 +209,7 @@ INT32U piDIOComm_sendCyclicTelegram(u8 devnum)
 	rcv_len = 3 * sizeof(u16) + i8uNumCounter[addr] * sizeof(u32);
 
 	ret = pibridge_req_io(addr, cmd, snd_buf, snd_len, in_buf, rcv_len);
-	if (ret) {
+	if (ret < 0) {
```
Comment (l1k) [inline piDIOComm.c:212]: Again, this needs to be `ret != rcv_len` so that too-short packets are treated as errors.

File: revpi_mio.c [hunk 11]
```diff
@@ -39,7 +39,7 @@ static int revpi_mio_cycle_dio(SDevice *dev, SMioDigitalRequestData *req_data,
 
 	ret = pibridge_req_io(dev->i8uAddress, IOP_TYP1_CMD_DATA,
 			      &req, sizeof(req), &resp, sizeof(resp));
-	if (ret) {
+	if (ret < 0) {
```
Comment (l1k) [inline revpi_mio.c:42]: `ret != sizeof(resp)`

File: revpi_mio.c [hunk 12]
```diff
@@ -63,7 +63,7 @@ static int revpi_mio_cycle_aio(SDevice *dev, SMioAnalogRequestData *req_data,
 	ret = pibridge_req_io(dev->i8uAddress, IOP_TYP1_CMD_DATA2,
 			      req_data, sizeof(*req_data) - compressed,
 			      &resp, sizeof(resp));
-	if (ret) {
+	if (ret < 0) {
```
Comment (l1k) [inline revpi_mio.c:66]: `ret != sizeof(resp)`

File: revpi_mio.c [hunk 13]
```diff
@@ -304,22 +304,22 @@ int revpi_mio_init(unsigned char devno)
 	/*dio*/
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_CFG,
 			      &conf->dio, sizeof(conf->dio), NULL, 0);
-	if (ret) {
+	if (ret < 0) {
```
Comment (l1k) [inline revpi_mio.c:307]: Unnecessary change.

File: revpi_mio.c [hunk 14]
```diff
@@ -304,22 +304,22 @@ int revpi_mio_init(unsigned char devno)
...
 		pr_err("talk with mio for conf dio err(devno:%d, ret:%d)\n",
 		       devno, ret);
 	}
 
 	/*aio in*/
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA4,
 			      &conf->aio_i, sizeof(conf->aio_i), NULL, 0);
-	if (ret)
+	if (ret < 0)
```
Comment (l1k) [inline revpi_mio.c:315]: Unnecessary change.

File: revpi_mio.c [hunk 15]
```diff
@@ -304,22 +304,22 @@ int revpi_mio_init(unsigned char devno)
...
 		pr_err("talk with mio for conf aio_i err(devno:%d, ret:%d)\n",
 		       devno, ret);
 
 	/*aio out*/
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA4,
 			      &conf->aio_o, sizeof(conf->aio_o), NULL, 0);
-	if (ret)
+	if (ret < 0)
```
Comment (l1k) [inline revpi_mio.c:322]: Unnecessary change.

File: RevPiDevice.c (hunk 1 above)
Comment (linosanfilippo-kunbus) [reply RevPiDevice.c:258]: AFAIK the only user of this function are we (KUNBUS). Do we really evaluate the returned Telegram? If not I would rather like to remove it from piCore_g.

File: piDIOComm.c (hunk 9 above)
Comment (linosanfilippo-kunbus) [reply piDIOComm.c:143]: Indeed

File: RevPiDevice.c (hunk 1 above)
Comment (l1k) [reply RevPiDevice.c:262]: I would be open to dropping the `KB_INTERN_IO_MSG` ioctl and then the `piCore_g.responseUserTel` becomes obsolete as well. But it's probably reasonable to do that separately and after coordination with the PO & team, so retaining backward compatibility for now (as you've done) seems like a good idea.

Comment (nbuchwitz) [reply RevPiDevice.c:262]: If I remember your explanations a few weeks ago this telegram type is (probably) is debug relic from our past? Then I'd be happy to drop this. Less code and more understood functionality hopefully increases the maintainability in the long run.

Comment (linosanfilippo-kunbus) [reply RevPiDevice.c:262]: Note, that the header of the response telegram does not contain any valid data. Should that also be adjusted to look like the header of a real response packet?

File: piAIOComm.c [hunk 16]
```diff
@@ -318,7 +318,9 @@ u32 piAIOComm_sendCyclicTelegram(u8 devnum)
 
 	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA, snd_buf,
 			      AIO_OUTPUT_DATA_LEN, rcv_buf, AIO_INPUT_DATA_LEN);
-	if (ret) {
+	if (ret != AIO_INPUT_DATA_LEN) {
+		if (ret > 0)
+			ret = -EIO;
 		pr_warn_ratelimited("addr %2d: io communication failed with AIO %d\n",
 			addr, ret);
 		return 1;
```
Comment (l1k) [inline piAIOComm.c:326]: Hm, I'm confused because the error path unconditionally returns `1`, so changing `ret` to `-EIO` only affects the warning message. Is that intentional? There's arguably not much of a benefit for the warning message to contain -EIO instead of a positive integer...

File: piDIOComm.c [hunk 17]
```diff
@@ -207,7 +207,9 @@ INT32U piDIOComm_sendCyclicTelegram(u8 devnum)
 	rcv_len = 3 * sizeof(u16) + i8uNumCounter[addr] * sizeof(u32);
 
 	ret = pibridge_req_io(addr, cmd, snd_buf, snd_len, in_buf, rcv_len);
-	if (ret) {
+	if (ret != rcv_len) {
+		if (ret > 0)
+			ret = -EIO;
 		pr_warn_ratelimited("addr %2d: io communication failed %d\n",
 			addr, ret);
 		return 1;
```
Comment (l1k) [inline piDIOComm.c:215]: Same here.

File: piAIOComm.c (hunk 16 above)
Comment (linosanfilippo-kunbus) [reply piAIOComm.c:326]: Yes, thats intentional. The only reason for the for the adjustment of "ret" was to have an error value in the following message instead of the number of received bytes

File: revpi_mio.c [hunk 18]
```diff
@@ -39,7 +39,9 @@ static int revpi_mio_cycle_dio(SDevice *dev, SMioDigitalRequestData *req_data,
 
 	ret = pibridge_req_io(dev->i8uAddress, IOP_TYP1_CMD_DATA,
 			      &req, sizeof(req), &resp, sizeof(resp));
-	if (ret) {
+	if (ret != sizeof(resp)) {
+		if (ret > 0)
+			ret = -EIO;
 		pr_err_ratelimited("talk with mio for dio data error(addr:%d, "
 				   "ret:%d)\n", dev->i8uAddress, ret);
 		return ret;
```
Comment (l1k) [inline revpi_mio.c:47]: In the MIO case, `ret` is passed up the call chain but it's only interpreted by <a href="https://github.com/RevolutionPi/piControl/blob/ab3b62822abed8933ffeaa1a7e6eef55768cb062/RevPiDevice.c#L152">revpi_dev_update_state()</a>, which only checks whether it's non-zero. So changing to `-EIO` doesn't actually change the outcome. Of course, there could be other reasons to change the return value (taste, making things foolproof, ...).

File: piIOComm.c [hunk 19]
```diff
@@ -311,7 +311,9 @@ INT32S piIoComm_sendRS485Tel(INT16U i16uCmd_p, INT8U i8uAddress_p,
 	ret = pibridge_req_gate_tmt(i8uAddress_p, i16uCmd_p, pi8uSendData_p,
 				    i8uSendDataLen_p, pi8uRecvData_p,
 				    rcvlen, timeout);
-	if (ret) {
+	if (ret != rcvlen) {
+		if (ret >= 0)
+			ret = -EIO;
```
Comment (l1k) [inline piIOComm.c:316]: I've looked at all the callers of `piIoComm_sendRS485Tel()` and they either only check that the return value is non-zero or they ignore the return value.

In particular, all the callers in <a href="https://github.com/RevolutionPi/piControl/blob/ab3b62822abed8933ffeaa1a7e6eef55768cb062/RevPiDevice.c">RevPiDevice.c</a> only check for a non-zero return value. Curiously, they all store the return value as an *unsigned* int even though `piIoComm_sendRS485Tel()` returns a signed int. Very odd.

The single caller in <a href="https://github.com/RevolutionPi/piControl/blob/ab3b62822abed8933ffeaa1a7e6eef55768cb062/PiBridgeMaster.c">PiBridgeMaster.c</a> stores the return value in `piCore_g.statusGateTel`. At least that's a signed integer. But the `KB_CONFIG_SEND` ioctl, which is the only one sending userspace gate telegrams, stores the return value in <a href="https://github.com/RevolutionPi/piControl/blob/ab3b62822abed8933ffeaa1a7e6eef55768cb062/piControlMain.c#L1501">status</a>, then unconditionally overwrites it with 0. So userspace never learns if sending the gate telegram failed. This looks like another bug, oh dear! :(

**Review by l1k — CHANGES_REQUESTED** [review]
Not really a change request, more like a "do you want to keep this unchanged despite my comments?"

**Comment by linosanfilippo-kunbus** [comment]
> Not really a change request, more like a "do you want to keep this unchanged despite my comments?"

If its the adjusted return values: that was by intention, knowing that the callers dont do anything sane with it. The only reason was to provide an error value for the following error message. So I do not want to change that.

**Review by l1k — APPROVED** [review]
